### PR TITLE
Add natspec

### PR DESCRIPTION
**Pull request by zodahu**
Natspec comments have been added to most `/src` contracts, except for
1. Some contracts in `/src/fees`, as they may be updated recently. For instance, not all functions in `FeeVerifier` have been provided with Natspec comments.
2. Some functions are defined in other protocols such as `MakerUtility` `openLockETHAndDraw`, `AaveV2FlashLoanCallback` `executeOperation`,  they don't include the @param Natspec comments, as they are already defined and documented within their respective protocols.

### CODE REVIEW

File: src/Router.sol [hunk 1]
```diff
@@ -122,21 +161,32 @@ contract Router is IRouter, EIP712, FeeVerifier {
...
 
+    /// @notice Resume `execute` and `executeWithSignature` by pauser
     function resume() external onlyPauser {
         paused = false;
         emit Resumed();
     }
 
-    /// @notice Execute logics through current user's agent. Create agent for user if not created.
+    /// @notice Execute arbitrary logics through the current user's agent. Creates an agent for users if not created.
+    ///         Charge fees based on the scenarios defined in the FeeVerifier contract, which calculates fees on-chain.
```
Comment (hihiben) [inline src/Router.sol:185]: Not sure if the indent is a correct format for spec parsing.

File: src/fees/FeeCalculatorBase.sol [hunk 2]
```diff
@@ -31,10 +44,16 @@ abstract contract FeeCalculatorBase {
         return amounts;
     }
 
+    /// @notice Calculate the amount with the fee included for the given amount
+    /// @param amount The amount to calculate the total with the fee included
```
Comment (hihiben) [inline src/fees/FeeCalculatorBase.sol:48]: Quite confusing on those fee calculating functions. Can we just simply say "The amount to be calculated"? All of those `amount` should be the amount without fee being included right?

File: src/fees/FeeVerifier.sol [hunk 3]
```diff
@@ -5,15 +5,23 @@ import {Ownable} from 'openzeppelin-contracts/contracts/access/Ownable.sol';
 import {IParam} from '../interfaces/IParam.sol';
 import {IFeeCalculator} from '../interfaces/IFeeCalculator.sol';
 
+/// @title Fee verifier
+/// @notice An abstruct contract that verifies and calculates fees on-chain
```
Comment (hihiben) [inline src/fees/FeeVerifier.sol:9]:
```suggestion
/// @notice An abstract contract that verifies and calculates fees on-chain
```

File: src/interfaces/IParam.sol [hunk 4]
```diff
@@ -2,40 +2,41 @@
...
     struct Logic {
-        address to;
-        bytes data;
-        Input[] inputs;
-        WrapMode wrapMode;
-        // Approve to another contract instead of `to` since some protocols use spender contract to pull tokens from user
-        address approveTo;
-        address callback;
+        address to; // The target address for the execution
+        bytes data; // Encoded function call data
```
Comment (hihiben) [inline src/interfaces/IParam.sol:15]:
```suggestion
        bytes data; // Encoded function calldata
```

File: src/utilities/MakerUtility.sol [hunk 5]
```diff
@@ -8,6 +8,8 @@ import {IDSProxy, IDSProxyRegistry} from '../interfaces/maker/IDSProxy.sol';
 import {IMakerGemJoin} from '../interfaces/maker/IMaker.sol';
 import {ApproveHelper} from '../libraries/ApproveHelper.sol';
 
+/// @title Maker utility contract
+/// @notice Perform additional actions after interacting with Maker
```
Comment (hihiben) [inline src/utilities/MakerUtility.sol:12]: Should be before? I thought user will call utility contract instead of calling Maker.

File: src/fees/FeeCalculatorBase.sol [hunk 6]
```diff
@@ -3,24 +3,37 @@ pragma solidity ^0.8.0;
...
 
+    /// @dev Initialize the router address and fee rate
     constructor(address router_, uint256 feeRate_) {
         router = router_;
         feeRate = feeRate_;
     }
 
+    /// @notice Calculate the fee for the given amount
+    /// @param amount The amount for which the fee needs to be calculated
+    /// @return The calculated fee
```
Comment (MussinaLin) [inline src/fees/FeeCalculatorBase.sol:29]: Because the param `amount` here is fee included, I think we should mention this. What do you think ?

```suggestion
    /// @param amount The amount which included the fee already
    /// @return The fee calculated from `amount`
```

File: src/interfaces/IParam.sol [hunk 7]
```diff
@@ -2,40 +2,41 @@
 pragma solidity ^0.8.0;
 
 interface IParam {
+    /// @notice LogicBatch is signed by a signer using EIP-712
     struct LogicBatch {
-        Logic[] logics;
-        Fee[] fees;
-        uint256 deadline;
+        Logic[] logics; // An array of logics to be executed
```
Comment (MussinaLin) [inline src/interfaces/IParam.sol:7]: Feel like this way is better, [ref](https://github.com/aave/aave-v3-core/blob/29ff9b9f89af7cd8255231bc5faf26c3ce0fb7ce/contracts/deployments/ReservesSetupHelper.sol#L31).  What do you think ?

```suggestion
        Logic[] logics; // An array of Logic struct to be executed
```

File: src/Router.sol (hunk 1 above)
Comment (zodahu) [reply src/Router.sol:185]: The parser simply encodes '\n' and indent:
image

The Solidity official documentation doesn't give a specific rule for this, and the referenced GitHub issue(https://github.com/ethereum/solidity/issues/11359) doesn't arrive at a solid conclusion either.

In our past development, we have chosen this particular style, which aligns with what is used by Curve, Gnosis, etc. This choice is also noted in this [ticket](https://app.asana.com/0/1110393615095724/1202201092359525/f).

File: src/interfaces/IParam.sol [hunk 8]
```diff
@@ -2,40 +2,41 @@
...
+        Fee[] fees; // An array of fees to be charged
```
Comment (MussinaLin) [inline src/interfaces/IParam.sol:8]: Same as Ln.7

File: src/fees/FeeCalculatorBase.sol (hunk 2 above)
Comment (zodahu) [reply src/fees/FeeCalculatorBase.sol:48]: Agreed. Let's use with the comment style as you've suggested.

File: src/utilities/MakerUtility.sol (hunk 5 above)
Comment (zodahu) [reply src/utilities/MakerUtility.sol:12]: This comment indicates that the transfers of DAI and CDP occur after interacting with Maker in each functions in Utility, However, I agree it might be confusing.

I suggest rewording it to:  Perform additional actions `when` interacting with Maker

File: src/fees/FeeCalculatorBase.sol (hunk 2 above)
Comment (zodahu) [reply src/fees/FeeCalculatorBase.sol:48]: Removed the previous comment.

The `@param amount` in both `calculateFee` functions are fee included. Do you think `calculateAmountWithFee` should remove those `included` terms?

File: src/interfaces/IParam.sol (hunk 7 above)
Comment (zodahu) [reply src/interfaces/IParam.sol]: Use `Logic`, `Input` and `Fee` structs in comment, thanks.

File: src/fees/FeeCalculatorBase.sol (hunk 6 above)
Comment (zodahu) [reply src/fees/FeeCalculatorBase.sol:29]: Fixed. I think that 'from amount' is not strictly necessary, feel free to comment if you strongly prefer it.

File: src/interfaces/IParam.sol (hunk 7 above)
Comment (zodahu) [reply src/interfaces/IParam.sol]: I think we should use plurals after `an array of`. What do you think?

File: src/fees/FeeCalculatorBase.sol (hunk 6 above)
Comment (MussinaLin) [reply src/fees/FeeCalculatorBase.sol:29]: looks good.

**Review by hihiben — APPROVED** [review]
LGTM!

File: src/interfaces/IParam.sol (hunk 7 above)
Comment (MussinaLin) [reply src/interfaces/IParam.sol]: What's your reason ?  I thought it means: an array of `xxx struct`.
